File: backend/app/customer_waf/schema/customer_waf.py
"""
Request / response shapes for the customer WAF routes.

Responses are a **deliberate projection** of what the WAF returns: the service
token (plaintext or ciphertext) never appears in any of them, WAF event
``raw_log`` stays out, and site TLS / auth configuration is not exposed.
"""

from datetime import datetime
from typing import List
from typing import Literal
from typing import Optional

from pydantic import BaseModel
from pydantic import ConfigDict
from pydantic import Field

# ── configuration ──────────────────────────────────────────────────────────


class WafInstance(BaseModel):
    """A customer's WAF as CoPilot stores it — everything except the token."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    customer_code: str
    name: str
    api_url: str
    token_prefix: str
    verify_tls: bool
    has_ca_cert: bool = False
    enabled: bool
    last_verified_at: Optional[datetime] = None
    last_verified_role: Optional[str] = None
    created_by: Optional[int] = None
    created_at: Optional[datetime] = None
    updated_by: Optional[int] = None
    updated_at: Optional[datetime] = None


class WafInstanceCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=100, description="Unique per customer, e.g. 'prod-eu'")
    api_url: str = Field(..., max_length=1024, description="WAF admin UI / API base URL, e.g. https://waf.example.com:8080")
    service_token: str = Field(..., description="A wafst_… service token from the WAF (write-only)")
    verify_tls: bool = True
    ca_cert_pem: Optional[str] = Field(None, max_length=65536, description="CA certificate(s) to verify the WAF against")
    enabled: bool = True


class WafInstanceUpdate(BaseModel):
    """All optional. A missing or blank ``service_token`` keeps the stored one; ``ca_cert_pem: ""`` removes the CA."""

    name: Optional[str] = Field(None, min_length=1, max_length=100)
    api_url: Optional[str] = Field(None, max_length=1024)
    service_token: Optional[str] = None
    verify_tls: Optional[bool] = None
    ca_cert_pem: Optional[str] = Field(None, max_length=65536)
    enabled: Optional[bool] = None


class WafCapabilities(BaseModel):
    can_read: bool = False
    can_block: bool = False
    can_manage_forwarders: bool = False


class WafVerifyResult(BaseModel):
    reachable: bool
    authenticated: bool
    reason: Optional[str] = Field(None, description="Stable failure code: unreachable, token_rejected, tls_error, …")
    detail: Optional[str] = None
    waf_user_email: Optional[str] = None
    waf_roles: List[str] = []
    capabilities: WafCapabilities = WafCapabilities()
    health: Optional[dict] = Field(None, description="WAF /health (db, redis) when reachable at this URL")


class WafInstancesResponse(BaseModel):
    instances: List[WafInstance]
    encryption_key_configured: bool = Field(..., description="False → the UI explains why WAFs can't be saved")
    success: bool
    message: str


class WafInstanceResponse(BaseModel):
    instance: WafInstance
    verification: Optional[WafVerifyResult] = None
    warnings: List[str] = []
    success: bool
    message: str


class WafVerifyResponse(BaseModel):
    instance: WafInstance
    verification: WafVerifyResult
    success: bool
    message: str


class WafDeleteResponse(BaseModel):
    success: bool
    message: str


# ── read views (projections of WAF responses) ──────────────────────────────


class WafSite(BaseModel):
    id: str
    name: str
    hostname: str
    upstream_url: str
    is_enabled: bool
    detection_mode: bool
    created_at: Optional[datetime] = None


class WafSitesResponse(BaseModel):
    sites: List[WafSite]
    success: bool
    message: str


class WafMatchedRule(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: Optional[str] = None
    msg: Optional[str] = None


class WafEvent(BaseModel):
    id: str
    timestamp: datetime
    transaction_id: str
    site_id: Optional[str] = None
    client_ip: str
    method: str
    uri: str
    host: str
    rule_id: Optional[str] = None
    action: str
    severity: Optional[str] = None
    anomaly_score: Optional[int] = None
    matched_rules: List[WafMatchedRule] = []
    geoip_country_code: Optional[str] = None
    geoip_country_name: Optional[str] = None
    geoip_city: Optional[str] = None


class WafEventsResponse(BaseModel):
    events: List[WafEvent]
    limit: int
    offset: int
    success: bool
    message: str


WafAction = Literal["blocked", "detected", "passed"]


class WafRuleCount(BaseModel):
    rule_id: Optional[str] = None
    count: int


class WafIpCount(BaseModel):
    client_ip: str
    count: int


class WafHourCount(BaseModel):
    hour: str
    count: int


class WafCategoryCount(BaseModel):
    category: str
    count: int


class WafCountryCount(BaseModel):
    country_code: Optional[str] = None
    country_name: Optional[str] = None
    count: int


class WafStats(BaseModel):
    total_1h: int
    blocked_1h: int
    detected_1h: int
    block_rate: float
    top_rules: List[WafRuleCount] = []
    top_ips: List[WafIpCount] = []
    requests_per_hour: List[WafHourCount] = []
    attack_categories: List[WafCategoryCount] = []
    top_countries: List[WafCountryCount] = []
    ingestion_lag_seconds: Optional[float] = None
    events_per_minute: Optional[float] = None
    caddy_healthy: Optional[bool] = None


class WafStatsResponse(BaseModel):
    stats: WafStats
    success: bool
    message: str


class WafThreatIntelEntry(BaseModel):
    ip_address: str
    block_count: int
    rule_hit_count: int
    total_events: int
    threat_score: int
    top_rules: List[WafRuleCount] = []
    country_code: Optional[str] = None
    country_name: Optional[str] = None
    first_seen_at: datetime
    last_seen_at: datetime
    is_blocked: bool


class WafThreatIntelSummary(BaseModel):
    total_ips: int
    last_run_at: Optional[datetime] = None
    window_days: Optional[int] = None
    min_score: Optional[int] = None


class WafThreatIntelResponse(BaseModel):
    summary: WafThreatIntelSummary
    entries: List[WafThreatIntelEntry]
    success: bool
    message: str

File: frontend/src/types/customer-waf.ts
// Mirrors backend/app/customer_waf/schema/customer_waf.py (#1165).
// The WAF's service token is write-only: no type here ever carries it back.

export interface CustomerWafInstance {
	id: number
	customer_code: string
	name: string
	api_url: string
	/** First characters of the token, e.g. "wafst_AbC12..." — the token itself is never returned. */
	token_prefix: string
	verify_tls: boolean
	has_ca_cert: boolean
	enabled: boolean
	last_verified_at: string | null
	/** Comma-joined WAF roles of the token's user at the last connection test. A hint, not authorisation. */
	last_verified_role: string | null
	created_by: number | null
	created_at: string | null
	updated_by: number | null
	updated_at: string | null
}

export interface CustomerWafPayload {
	name: string
	api_url: string
	/** Required on create. On update, blank keeps the stored token. */
	service_token?: string
	verify_tls: boolean
	ca_cert_pem?: string | null
	enabled: boolean
}

export interface CustomerWafCapabilities {
	can_read: boolean
	can_block: boolean
	can_manage_forwarders: boolean
}

/** Stable failure codes from the backend; see app/customer_waf/utils/universal.py. */
export type CustomerWafFailureReason =
	| "unreachable"
	| "tls_error"
	| "token_rejected"
	| "insufficient_role"
	| "not_a_waf"
	| "not_found"
	| "bad_response"
	| "upstream_error"
	| "disabled"
	| "token_crypto"
	| "invalid_target"
	| "blocked_by_waf_rule"
	| "not_blocked"

export interface CustomerWafVerification {
	reachable: boolean
	authenticated: boolean
	reason: CustomerWafFailureReason | null
	detail: string | null
	waf_user_email: string | null
	waf_roles: string[]
	capabilities: CustomerWafCapabilities
	health: { status?: string; db?: string; redis?: string } | null
}

export interface CustomerWafSite {
	id: string
	name: string
	hostname: string
	upstream_url: string
	is_enabled: boolean
	detection_mode: boolean
	created_at: string | null
}

export type CustomerWafAction = "blocked" | "detected" | "passed"

export interface CustomerWafEvent {
	id: string
	timestamp: string
	transaction_id: string
	site_id: string | null
	client_ip: string
	method: string
	uri: string
	host: string
	rule_id: string | null
	action: CustomerWafAction | string
	severity: string | null
	anomaly_score: number | null
	matched_rules: { id: string | null; msg: string | null }[]
	geoip_country_code: string | null
	geoip_country_name: string | null
	geoip_city: string | null
}

export interface CustomerWafEventsQuery {
	action?: CustomerWafAction | null
	severity?: string | null
	client_ip?: string | null
	rule_id?: string | null
	site_id?: string | null
	start_time?: string | null
	end_time?: string | null
	limit?: number
	offset?: number
}

export interface CustomerWafStats {
	total_1h: number
	blocked_1h: number
	detected_1h: number
	block_rate: number
	top_rules: { rule_id: string | null; count: number }[]
	top_ips: { client_ip: string; count: number }[]
	requests_per_hour: { hour: string; count: number }[]
	attack_categories: { category: string; count: number }[]
	top_countries: { country_code: string | null; country_name: string | null; count: number }[]
	ingestion_lag_seconds: number | null
	events_per_minute: number | null
	caddy_healthy: boolean | null
}

export interface CustomerWafThreatIntelEntry {
	ip_address: string
	block_count: number
	rule_hit_count: number
	total_events: number
	threat_score: number
	top_rules: { rule_id: string | null; count: number }[]
	country_code: string | null
	country_name: string | null
	first_seen_at: string
	last_seen_at: string
	is_blocked: boolean
}

export interface CustomerWafThreatIntelSummary {
	total_ips: number
	last_run_at: string | null
	window_days: number | null
	min_score: number | null
}

export interface CustomerWafBlock {
	target: string
	rule_uuid: string
	rule_id: number
	name: string
	description: string
	enabled: boolean
	created_by_copilot: boolean
	created_at: string | null
}

export type CustomerWafBlockAction = "created" | "reenabled" | "already_blocked" | "blocked_by_waf_rule"

export interface CustomerWafBlockPayload {
	target: string
	reason: string
	alert_id?: number | null
	case_id?: number | null
}
